types/mempool/mempool.go

```diff
@@ -0,0 +1,254 @@
+package mempool
+
+import (
+	"fmt"
+	"math"
+
+	huandu "github.com/huandu/skiplist"
+
+	"github.com/cosmos/cosmos-sdk/types"
+	"github.com/cosmos/cosmos-sdk/x/auth/signing"
+)
+
+// Tx we define an app-side mempool transaction interface that is as
+// minimal as possible, only requiring applications to define the size of the
+// transaction to be used when reaping and getting the transaction itself.
+// Interface type casting can be used in the actual app-side mempool implementation.
+type Tx interface {
+	types.Tx
+
+	// Size returns the size of the transaction in bytes.
+	Size() int64
+}
+
+type Mempool interface {
+	// Insert attempts to insert a Tx into the app-side mempool returning
+	// an error upon failure.
+	Insert(types.Context, Tx) error
+
+	// Select returns the next set of available transactions from the app-side
+	// mempool, up to maxBytes or until the mempool is empty. The application can
+	// decide to return transactions from its own mempool, from the incoming
+	// txs, or some combination of both.
+	Select(txs [][]byte, maxBytes int64) ([]Tx, error)
+
+	// CountTx returns the number of transactions currently in the mempool.
+	CountTx() int
+
+	// Remove attempts to remove a transaction from the mempool, returning an error
+	// upon failure.
+	Remove(Tx) error
+}
+
+type ErrTxNotFound struct {
+	error
+}
+
+type Factory func() Mempool
+
+var _ Mempool = (*defaultMempool)(nil)
```

Comment:
nit: can we keep this file minimal and move the default impl to another file like `default.go` or `default_mempool.go`?